### PR TITLE
[FEAT] Detect equi-join conditions in a blocking rule to count the number of comparisons without needing to perform the join 

### LINKED ISSUE #1376: [FEAT] Detect equi-join conditions in a blocking rule to count the number of comparisons without needing to perform the join
We currently use a very inefficient algorithm to [count_num_comparisons_from_blocking_rule ](https://moj-analytical-services.github.io/splink/linkerexp.html?h=count_#splink.linker.Linker.count_num_comparisons_from_blocking_rule)

There's an obvious improvement, but it requires analysis of the sql to distinguish:
-  equi-join conditions in a blocking rule (e.g. `l.name = r.name`, or `substr(l.dob,1,2) = substr(r.dob,1,2)`, `l.surname = r.first_name`
-  non equi-join conditions (filters) like `jaro(l.name, r.name) > 0.9` 

Here's a list of test cases, all of which need to be supported:
```
blocking_rules = [
    "l.first_name = r.first_name",
    "l.first_name = r.first_name AND l.surname = r.surname",
    "substr(l.first_name,2,3) = substr(r.first_name,3,4)",
    "substr(l.first_name,1,1) = substr(r.surname,1,1) and l.dob = r.dob",
    "l.first_name = r.first_name and levenshtein(l.dob, r.dob) < 2",
    "substr(l.first_name,2,3) = substr(r.first_name,3,4) and levenshtein(l.dob, r.dob) < 2",
    "l.first_name = r.first_name OR l.surname = r.surname",
    "l.dob = r.dob and substr(l.first_name,2,3) = substr(r.first_name,3,4)",
]
```
At the moment, to count the number of comparisons from a blocking rule, we run the relevant SQL join (usually a self-join) with the blocking rule as the ‘inner join’ condition.

That is, the SQL engine must generate the pairwise comparisons in order to count them.

But, in the case that the blocking rule is a equi-join condition, this is unnecessary.  

For example, if the  blocked rule was on first name, then instead of generating all John -> John comparisons, you can just say ‘there are four johns, so n(n-1)/2 comparisons among them’ - i.e. you can perform an aggregation on the input nodes, there's no need to create the comparisons (edges)

To do this we will need to detect which conditions are 'hashable' (spark terminology) i.e. which are 'equality conditions' between information in the left and right table:

So for example:
`l.name = r.name` : ✅ 
`l.name = r.name and l.dob = r.dob`✅ 
`substr(l.name, 1,3) = substr(r.name, 1,3)` ✅ 
`substr(l.name, 1,3) = substr(r.name, 2,4)` ✅ 
`jaro(l.name, r.name) > 0.9` ❌

This could also be useful:
* to alert users when their blocking rules are non-hashable and therefore inefficient
* as the foundation to auto-generate a list of possible blocking rules
* possible to run a check prior to actually running predict() that the number of comparisons generated is not enormous

Relevant issues:
https://github.com/moj-analytical-services/splink/issues/1188
https://github.com/moj-analytical-services/splink/issues/1171

Worth noting i had a spark-only solution for this in Splink 2, but it's quite a brittle approach:
https://github.com/moj-analytical-services/splink/blob/9c9984892ed6306a9e06f0e599a4ee5abdf0939c/splink/analyse_blocking_rule.py

### PR DESCRIPTION
**Pull request by RobinL**
### Is your Pull Request linked to an existing Issue or Pull Request?
Will close #1376

Will also set the groundwork for a future method like `linker.suggest_blocking_rules(target_comparisons=x)`

### Give a brief description for the solution you have provided

See the #1376 issue for more detail:

- Use SQLGlot to extract the equi-join conditions
- Write a SQL query that uses these conditions to 'forecast' the [count_num_comparisons_from_blocking_rule](https://moj-analytical-services.github.io/splink/linkerexp.html?h=count_#splink.linker.Linker.count_num_comparisons_from_blocking_rule)

#### To do:

- [x] Add tests
- [ ] Add tests for different dialects, making sure to use dialect-specific syntax in the blocking rule 
- [ ] Manually try and actual example in Spark to double check spark-specific syntax works fine
- [x] Make sure it works for all link types
- [x] Move code to `analyse_blocking.py`

### PR Checklist

- [ ] Added documentation for changes.   👈 Not yet, since it's currently a private (deliberately undocumented) feature
- [x] Added tests (if appropriate)
- [x] Made changes based off the latest version of Splink
- [x] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint.html)

### CODE REVIEW

**Comment by RobinL** [comment]
As expected, the new methodology is dramatically quicker, especially on loose blocking conditions (which are the ones which are most problematic and we most need to be able to analyse easily):

```
df = pd.read_parquet(
    "/Users/robinlinacre/Documents/data_linking/splink/synthetic_1m_clean.parquet"
)



blocking_rules = [
    "l.first_name = r.first_name",
    "l.first_name = r.first_name AND l.surname = r.surname",
    "substr(l.first_name,2,3) = substr(r.first_name,3,4)",
    "substr(l.first_name,1,3) = substr(r.first_name,1,3)",

]

settings = {"link_type": "dedupe_only"}

linker = DuckDBLinker(df, settings)
for br in blocking_rules:
    print("-------")
    print(br)

    start_time = datetime.datetime.now()
    print(linker.count_num_comparisons_from_blocking_rule_2(br))
    end_time = datetime.datetime.now()
    time_new = end_time - start_time
    print("Time taken for count_num_comparisons_from_blocking_rule_2: ", time_new)

    start_time = datetime.datetime.now()
    print(linker.count_num_comparisons_from_blocking_rule(br))
    end_time = datetime.datetime.now()
    time_old = end_time - start_time
    print(
        "Time taken for count_num_comparisons_from_blocking_rule: ",
        time_old,
    )

    print(f"Speed multipler: {(time_old/time_new):,.1f}")

```

```
-------
l.first_name = r.first_name
5146097324.0
Time taken for count_num_comparisons_from_blocking_rule_2:  0:00:00.166095
2572496060
Time taken for count_num_comparisons_from_blocking_rule:  0:00:30.821882
Speed multipler: 185.6
-------
l.first_name = r.first_name AND l.surname = r.surname
12628073.0
Time taken for count_num_comparisons_from_blocking_rule_2:  0:00:00.182770
5807175
Time taken for count_num_comparisons_from_blocking_rule:  0:00:00.435371
Speed multipler: 2.4
-------
substr(l.first_name,2,3) = substr(r.first_name,3,4)
455003494.0
Time taken for count_num_comparisons_from_blocking_rule_2:  0:00:00.074847
228368393
Time taken for count_num_comparisons_from_blocking_rule:  0:00:06.597419
Speed multipler: 88.1
-------
```

**Comment by RobinL** [comment]
Spark test script:
<details>

```python

from pyspark.context import SparkConf, SparkContext
from pyspark.sql import SparkSession

from splink.spark.linker import SparkLinker

conf = SparkConf()
conf.set("spark.driver.memory", "12g")
conf.set("spark.sql.shuffle.partitions", "12")

sc = SparkContext.getOrCreate(conf=conf)
sc.setCheckpointDir("tmp_checkpoints/")
spark = SparkSession(sc)

settings = {
    "link_type": "dedupe_only",
}


df = spark.read.csv("./tests/datasets/fake_1000_from_splink_demos.csv", header=True)

df = df.withColumn("hi THERE", "email")

linker = SparkLinker(df, input_table_aliases="fake_data_1")
br = "l.`first_name` = r.`first_name`"
linker._count_num_comparisons_from_blocking_rule_pre_filter_conditions()

```
</details>

File: splink/blocking.py
Comment (ThomasHepworth) [inline splink/blocking.py:61]: I'll start a formal review shortly, but can you sort out the poetry dependencies if sqlglot 11.4.1 is required for this PR?

I would imagine we can pin the version to 11.4.1 without any issues.

Comment (RobinL) [reply splink/blocking.py:61]: Good point. I haven't tried any other version actually, I just was m aking a note for any future readers of the code.  Definitely worth checking a couple of other versions in the range currently specified

File: splink/analyse_blocking.py
Comment (ThomasHepworth) [inline splink/analyse_blocking.py:147]: FYI, you'll be able to use [this new conversion function](https://github.com/moj-analytical-services/splink/blob/33fe61f646f8c2ec959ad0f393969bcdc46744c7/splink/blocking.py#L16) once some of my BR work has been merged in.

File: splink/analyse_blocking.py
Comment (ThomasHepworth) [inline splink/analyse_blocking.py:171]: Whether you want to is a different matter, but this can be simplified, making this section of the code less verbose and overbearing:
```suggestion
    l_cols_gb, r_cols_gb = zip(*join_conditions)
    using = [f"key_{i}" for i in range(len(join_conditions))]
    l_cols_sel = ", ".join([f"{l_c} as {u}" for l_c, u in zip(l_cols_gb, using)])
    r_cols_sel = ", ".join([f"{r_c} as {u}" for r_c, u in zip(r_cols_gb, using)])
    l_cols_gb = ", ".join(l_cols_gb)
    r_cols_gb = ", ".join(r_cols_gb)
    using = ", ".join(using)
```

File: splink/analyse_blocking.py
Comment (ThomasHepworth) [inline splink/analyse_blocking.py:164]: If you'd prefer to keep the initial for loop, then
```suggestion
        l_cols_gb.append(l_key)
        r_cols_gb.append(r_key)
```

File: splink/analyse_blocking.py
Comment (ThomasHepworth) [reply splink/analyse_blocking.py]: That's fair! I find it slightly easier to read, but there is no right answer here.

Marking as resolved.

File: splink/blocking.py
Comment (ThomasHepworth) [inline splink/blocking.py:75]: Just a general Q - is this better than using:
`c.args["table"] = None` here?

File: splink/blocking.py
Comment (ThomasHepworth) [inline splink/blocking.py:95]: You may well have noticed, but sqlglot also takes care of duplicated equalities... i.e.:
`"l.first_name = r.first_name"` yields the same `_join_conditions` as `"l.first_name = r.first_name and l.first_name = r.first_name"`

File: splink/blocking.py
Comment (ThomasHepworth) [inline splink/blocking.py:23]: A more general note on inheritance of the dialect... we may want to think about a more effective strategy for ensuring both our BRs and CLs inherit a sql dialect where possible.

File: splink/blocking.py
Comment (ThomasHepworth) [inline splink/blocking.py:98]: I think `filter_condition` is more interpretable for users than `non-equi-join`

File: splink/analyse_blocking.py
Comment (ThomasHepworth) [inline splink/analyse_blocking.py:199]: This appears to be broken for "stacked" dataframes at present - i.e. a link job where we input a single dataframe with a column detailing which records belongs to which dataset.

```
from splink.duckdb.linker import DuckDBLinker
import pandas as pd
df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
df['source_dataset'] = pd.Series(['a'] * 500 + ['b'] * 500, name='source_dataset')

linker = DuckDBLinker(
    df,
    settings_dict = {
    "link_type": "link_only",
    }
)

linker._initialise_df_concat(True).as_pandas_dataframe()
linker._two_dataset_link_only  # False

br = "levenshtein(first_name, 3)"
display(
    linker._count_num_comparisons_from_blocking_rule_pre_filter_conditions(br)
) # expecting 25k
```

**Comment by ThomasHepworth** [comment]
I'm essentially happy to approve this once the "stacked dataframe" problem has been resolved. That may require its own PR, because some of the logic may need to be rewritten.

Apologies, I meant to clean up some of that code a while back, but never found the time.

File: splink/blocking.py
Comment (RobinL) [reply splink/blocking.py:75]: Hmm, I'm not sure.  I guess ideally we'd want to be consistent with the rest of the Sqlglot codebase

File: splink/blocking.py
Comment (RobinL) [reply splink/blocking.py:95]: Nice, I had not noticed

File: splink/linker.py
Comment (RobinL) [reply splink/linker.py:2747]: Good point, I will check

File: splink/analyse_blocking.py
Comment (RobinL) [reply splink/analyse_blocking.py:199]: Ouch, good spot

File: splink/blocking.py
Comment (RobinL) [reply splink/blocking.py:61]: [Done](https://github.com/moj-analytical-services/splink/pull/1388/commits/4b992da1f9d9ad3ab663dac91389a8fdfc25d535)

File: splink/blocking.py
Comment (RobinL) [reply splink/blocking.py:75]:
```
sqlglot.parse_one("l.first_name").args
```
results in 
```
{'this': (IDENTIFIER this: first_name, quoted: False),
 'table': (IDENTIFIER this: l, quoted: False),
 'db': None,
 'catalog': None}
```

```
sqlglot.parse_one("first_name").args
```
results in 
```
{'this': (IDENTIFIER this: first_name, quoted: False)}
```

i.e. there is no table key, meaning that `c.args["table"]` reflects how sqlglot would represent column without a table prefix

File: splink/blocking.py
Comment (RobinL) [reply splink/blocking.py:93]: I [deleted it](https://github.com/moj-analytical-services/splink/pull/1388/commits/e8f3f3c71d6d7cccb4b5b48f9eeda2af52e6c57f)

File: splink/blocking.py
Comment (RobinL) [reply splink/blocking.py:23]: Agreed

File: splink/linker.py
Comment (RobinL) [reply splink/linker.py]: [done](https://github.com/moj-analytical-services/splink/pull/1388/commits/394d5f20ada7d4fd966ea73478534f6d5487c14b)

File: splink/analyse_blocking.py
Comment (RobinL) [reply splink/analyse_blocking.py:199]: I think this is broken more widely.  Consider running the following on `master`:
```
from splink.duckdb.linker import DuckDBLinker
import pandas as pd
df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
df['source_dataset'] = pd.Series(['a'] * 500 + ['b'] * 500, name='source_dataset')

linker = DuckDBLinker(
    df,
    settings_dict = {
    "link_type": "link_and_dedupe"
    }
)

import logging 
logging.getLogger("splink").setLevel(1)
linker._initialise_df_concat_with_tf(True).as_pandas_dataframe()
```

You still end up with a column called `__splink_source_dataset`

Comment (RobinL) [reply splink/analyse_blocking.py:199]: Oh wait - I don't think this is actually broken.    It's a little difficult to explain, but let me try:

- `link_only` works in the general _n_ dataset case, including `n==2`
- In the general case, a self-join of the concatenated datasets is used to generate comparisons, filtering out records where `source_dataset_l = source_dataset_r`
- There is a specific optimisation where splink observes two input datasets that allows an inner join of the raw (non concatenated) input datasets
- This optimisation is not used if you pass Splink a single dataframe that implicitly has two source datasets
- `1,000,000` is therefore the correct result from your code example

Comment (RobinL) [reply splink/analyse_blocking.py:199]: Separately, there is something a bit weird here for which I will raise a different issue.  Consider running the following on `master`:
```
from splink.duckdb.linker import DuckDBLinker
import pandas as pd
df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
df['source_dataset'] = pd.Series(['a'] * 500 + ['b'] * 500, name='source_dataset')

linker = DuckDBLinker(
    df,
    settings_dict = {
    "link_type": "link_and_dedupe"
    }
)

import logging 
logging.getLogger("splink").setLevel(1)
linker._initialise_df_concat_with_tf(True).as_pandas_dataframe()
```

You still end up with a column called `__splink_source_dataset`, which seems unnecessary

I will raise an issue - I think this probably needs to be fixed separately

Comment (ThomasHepworth) [reply splink/analyse_blocking.py:199]: Yeah, it's definitely unnecessary.

The fix that I put in place was merely intended as a bandage before a more permanent solution was implemented.

Apologies, I should've made an issue at the time.

**Review by ThomasHepworth — APPROVED** [review]
Thanks Robin, this is great!

Some caveats to be dealt with in a later PR.

**Pull request merged**
